jsonComposer.go

```diff
@@ -0,0 +1,218 @@
+// Copyright 2018-2019 Couchbase, Inc. All rights reserved.
+
+package gojsonsm
+
+import (
+	"fmt"
+)
+
+type jsonObjComposer struct {
+	// should have enough length
+	body []byte
+	// cursor
+	pos int
+	// if the newly composed json has atleast one field
+	atLeastOneFieldLeft bool
+}
+
+func (composer *jsonObjComposer) Write(data []byte) {
+	copy(composer.body[composer.pos:composer.pos+len(data)], data)
+	composer.pos += len(data)
+}
+
+func (composer *jsonObjComposer) Commit(stepBackAndClose bool) (int, bool) {
+	// close the JSON object after stepping back by one position explicity if requested
+	if stepBackAndClose && composer.pos-1 >= 0 {
+		if composer.body[composer.pos-1] != '{' {
+			composer.pos--
+		}
+		composer.body[composer.pos] = '}'
+		composer.pos++
+	}
+
+	// if the body is anything other than {}, we have atleast one field inside it
+	composer.atLeastOneFieldLeft = composer.pos > 2
+
+	return composer.pos, composer.atLeastOneFieldLeft
+}
+
+// Given a byte encoded json object - "src", a list of keys of items to remove from src - "remove",
+// the function removes the items from "src" and places them in "removed" and the remaining json object is stored in "dst"
+// returns (final length of dst, number of items removed, if there are any items left in dst at the end, error)
+// Caller has the ability to allocate memory for "dst" and "removed". If nil is passed, only then memory is allocated.
+func MatchAndRemoveItemsFromJsonObject(src []byte, remove []string, dst []byte, removed map[string][]byte) (int, int, bool, error) {
+	var removedLen int
+	if removed == nil {
+		removed = make(map[string][]byte)
+	}
+
+	if dst == nil {
+		dst = make([]byte, len(src))
+	}
+
```

Comment:
Do `src` variable check and error out if there are issues before instantiating new objects of `jsonObjComposer` and `jsonTokenizer`

Reply:
If users pass in a `src` JSON that is not an object, but purely just a string, etc. 
The logic below with depth will malfunction
And I don't think you will return "invalid JSON object" in that case, which this function should

Reply:
Do you mean to check the 0th and n-1th index for `{` and `}` ?
OR
Do you want the function to pass through the entire input and check in the beginning and check the entire input to see if it is a valid JSON object and if not return an error?
This will then become a 2-pass algorithm and will not be a 1-pass or near 1-pass algorithm anymore.

Reply:
Assuming you meant (1), I have updated.